src/libraries/System.Private.CoreLib/src/System/Text/Ascii.cs

```diff
@@ -14,56 +16,215 @@ public static partial class Ascii
+                            return false;
+                        }
+
+                        searchSpace = ref Unsafe.Add(ref searchSpace, 1);
+                    }
+
+                    return true;
+                }
+
+                // If vectorization isn't supported, process 16 bytes at a time.
+                if (!Vector128.IsHardwareAccelerated && length > 2 * elementsPerUlong)
+                {
+                    ref T finalStart = ref Unsafe.Subtract(ref searchSpaceEnd, 2 * elementsPerUlong);
+
+                    do
+                    {
+                        if (!AllCharsInUInt64AreAscii<T>(
+                            Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<T, byte>(ref searchSpace)) |
+                            Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref Unsafe.As<T, byte>(ref searchSpace), sizeof(ulong)))))
+                        {
+                            return false;
+                        }
+
+                        searchSpace = ref Unsafe.Add(ref searchSpace, 2 * elementsPerUlong);
+                    }
+                    while (Unsafe.IsAddressLessThan(ref searchSpace, ref finalStart));
+
+                    searchSpace = ref finalStart;
+                }
+
+                // Process the last [8, 16] bytes.
+                return AllCharsInUInt64AreAscii<T>(
+                    Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<T, byte>(ref searchSpace)) |
+                    Unsafe.ReadUnaligned<ulong>(ref Unsafe.Subtract(ref Unsafe.As<T, byte>(ref searchSpaceEnd), sizeof(ulong))));
+            }
+
+            // Process inputs with lengths [16, 32] bytes.
+            if (length <= 2 * Vector128<T>.Count)
+            {
+                return AllCharsInVectorAreAscii(
+                    Vector128.LoadUnsafe(ref searchSpace) |
+                    Vector128.LoadUnsafe(ref Unsafe.Subtract(ref searchSpaceEnd, (nuint)Vector128<T>.Count)));
+            }
+
+            if (Vector256.IsHardwareAccelerated)
+            {
+                // Process inputs with lengths [33, 64] bytes.
+                if (length <= 2 * Vector256<T>.Count)
+                {
+                    return AllCharsInVectorAreAscii(
+                        Vector256.LoadUnsafe(ref searchSpace) |
+                        Vector256.LoadUnsafe(ref Unsafe.Subtract(ref searchSpaceEnd, (nuint)Vector256<T>.Count)));
+                }
+
+                // Process long inputs 128 bytes at a time.
+                if (length > 4 * Vector256<T>.Count)
+                {
+                    // Process the first 128 bytes.
+                    if (!AllCharsInVectorAreAscii(
+                        Vector256.LoadUnsafe(ref searchSpace) |
+                        Vector256.LoadUnsafe(ref searchSpace, (nuint)Vector256<T>.Count) |
+                        Vector256.LoadUnsafe(ref searchSpace, 2 * (nuint)Vector256<T>.Count) |
+                        Vector256.LoadUnsafe(ref searchSpace, 3 * (nuint)Vector256<T>.Count)))
```

Comment:
How have we decided that 4 vectors is the right number? i.e. why not 3 or 5?

Reply:
CPUs generally support parallel dispatch of 1, 2, or 4. It is rarely 3, 5, or another number.

Reply:
If we do 4 here and the CPU only supports parallel dispatch of 1 or 2, are we then penalizing the implementation on that CPU or is it no worse than just having done one at a time?

Reply:
It won't penalize at all, it will effectively just execute 2 and then 2 again after the first two have completed. It "can" be less optimal than an alternative implementation, however.

Most instructions have effectively three key parts:
1. Latency - simply how many cycles a given instruction takes to execute.
2. Ports Used - which execution handler the instruction runs on
3. Reciprocal Throughput - how many cycles until another instruction can utilize the port(s)

Loads and stores are a bit more special than most as they also factor in a load-store dispatch unit and may be able to sustain more, depending on various factors.

Since you have these 3 factors to consider, there are some scenarios where interleaving instructions correctly can lead to better throughput. Say you have `2` ports available to execute `or` and those are distinct from the `1` port available to emit a `shuffle`. In that case, you may see better performance by doing `or, or, shuffle, or, or, shuffle` as compared to `or, or, or, or, shuffle, shuffle`.

Reply:
Got it. Thanks.

Reply:
In this case, we don't really have any alternative instructions we could interleave so it should be good as it is.

Reply:
With how the length checks are structured, powers of two are the simplest option.
If we want to read 5 vectors in the inner loop, then we have to add another length check before, at which point we may as well go up to 8 vectors at a time.

Results for different numbers of vectors per iteration:
|            Method | Toolchain | Length |      Mean |    Error | Code Size |
|------------------ |---------- |------- |----------:|---------:|----------:|
| AsciiIsValid_Byte |      pr-2 |  10000 |  77.48 ns | 0.292 ns |     539 B |
| AsciiIsValid_Byte |      pr-3 |  10000 |  64.17 ns | 0.261 ns |     544 B |
| AsciiIsValid_Byte |      **pr-4** |  10000 |  **54.67 ns** | 0.164 ns |     542 B |
| AsciiIsValid_Byte |      pr-5 |  10000 |  58.33 ns | 0.246 ns |     632 B |
| AsciiIsValid_Byte |      pr-6 |  10000 |  60.15 ns | 0.226 ns |     644 B |
| AsciiIsValid_Byte |      pr-7 |  10000 |  41.98 ns | 0.173 ns |     652 B |
| AsciiIsValid_Byte |      pr-8 |  10000 |  41.32 ns | 0.049 ns |     660 B |
|                   |           |        |           |          |           |
| AsciiIsValid_Char |      pr-2 |  10000 | 149.23 ns | 0.554 ns |     507 B |
| AsciiIsValid_Char |      pr-3 |  10000 | 118.97 ns | 0.339 ns |     512 B |
| AsciiIsValid_Char |      **pr-4** |  10000 | **102.44 ns** | 0.327 ns |     520 B |
| AsciiIsValid_Char |      pr-5 |  10000 |  84.25 ns | 0.394 ns |     614 B |
| AsciiIsValid_Char |      pr-6 |  10000 |  81.14 ns | 0.049 ns |     622 B |
| AsciiIsValid_Char |      pr-7 |  10000 |  76.76 ns | 0.063 ns |     630 B |
| AsciiIsValid_Char |      pr-8 |  10000 |  77.50 ns | 0.329 ns |     638 B |

You can get a bit more out of it by doing even more per iteration, at the cost of more code.

---

Though all of these are so fast that if loads aren't hitting L1/L2 cache, other details don't matter as much anymore.

|            Method | Toolchain |   Length |         Mean |       Error | Ratio |
|------------------ |---------- |--------- |-------------:|------------:|------:|
| AsciiIsValid_Byte |      main |   100000 |   1,716.5 ns |    30.78 ns |  1.00 |
| AsciiIsValid_Byte |      pr-1 |   100000 |   1,197.2 ns |     7.28 ns |  0.70 |
| AsciiIsValid_Byte |      pr-2 |   100000 |     949.4 ns |     4.84 ns |  0.55 |
| AsciiIsValid_Byte |      pr-4 |   100000 |     702.9 ns |     6.67 ns |  0.41 |
|                   |           |          |              |             |       |
| AsciiIsValid_Byte |      main | 10000000 | 372,828.3 ns | 5,904.73 ns |  1.00 |
| AsciiIsValid_Byte |      pr-1 | 10000000 | 359,243.2 ns | 2,970.70 ns |  0.96 |
| AsciiIsValid_Byte |      pr-2 | 10000000 | 351,752.7 ns | 3,672.46 ns |  0.94 |
| AsciiIsValid_Byte |      pr-4 | 10000000 | 344,010.9 ns | 2,840.98 ns |  0.92 |